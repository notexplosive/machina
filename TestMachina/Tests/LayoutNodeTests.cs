using ApprovalTests;
using ApprovalTests.Reporters;
using Machina.Data.Layout;
using Microsoft.Xna.Framework;
using TestMachina.Utility;
using Xunit;

namespace TestMachina.Tests
{

    public class LayoutNodeTests
    {
        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void linear_layout_test()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 100), LayoutStyle.Empty,
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.Leaf("item-2", LayoutSize.StretchedHorizontally(20)),
                LayoutNode.Leaf("item-3", LayoutSize.StretchedBoth())
            );

            var layoutResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(layoutResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void linear_layout_test_with_margin()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 100), new LayoutStyle(new Point(10, 5), 0),
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.Leaf("item-2", LayoutSize.StretchedHorizontally(20)),
                LayoutNode.Leaf("item-3", LayoutSize.StretchedBoth())
            );

            var layoutResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(layoutResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void linear_layout_test_with_padding()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 100), new LayoutStyle(Point.Zero, 5),
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.Leaf("item-2", LayoutSize.StretchedHorizontally(20)),
                LayoutNode.Leaf("item-3", LayoutSize.StretchedBoth())
            );

            var layoutResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(layoutResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void linear_layout_test_with_margin_and_padding()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 100), new LayoutStyle(padding: 5, margin: new Point(3, 6)),
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.Leaf("item-2", LayoutSize.StretchedHorizontally(20)),
                LayoutNode.Leaf("item-3", LayoutSize.StretchedBoth())
            );

            var layoutResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(layoutResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void nested_layout_test()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 100), LayoutStyle.Empty,
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.HorizontalParent("item-2", LayoutSize.StretchedHorizontally(20), new LayoutStyle(new Point(2, 3), 0),
                        LayoutNode.Leaf("item-2a", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2b", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2c", LayoutSize.StretchedHorizontally(10))
                    ),
                LayoutNode.VerticalParent("item-3", LayoutSize.StretchedBoth(), new LayoutStyle(new Point(0, 2), 3),
                        LayoutNode.Leaf("item-3a", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3b", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3c", LayoutSize.StretchedHorizontally(10))
                    )
            );

            var layoutResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(layoutResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void resize_and_rebake_test()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 50), LayoutStyle.Empty,
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.HorizontalParent("item-2", LayoutSize.StretchedHorizontally(20), new LayoutStyle(new Point(2, 3), 0),
                        LayoutNode.Leaf("item-2a", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2b", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2c", LayoutSize.StretchedHorizontally(10))
                    ),
                LayoutNode.VerticalParent("item-3", LayoutSize.StretchedBoth(), new LayoutStyle(new Point(5, 2), 1),
                        LayoutNode.Leaf("item-3a", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3b", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3c", LayoutSize.StretchedHorizontally(10))
                    )
            );

            var resizedLayout = layout.GetResized(new Point(60, 60));

            var firstBakeResult = layout.Bake();
            var secondBakeResult = resizedLayout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(firstBakeResult) + "\n\n" + LayoutNodeUtils.DrawResult(secondBakeResult));
        }

        [Fact]
        public void resize_performance_test()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 50), LayoutStyle.Empty,
                LayoutNode.Leaf("item-1", LayoutSize.StretchedHorizontally(10)),
                LayoutNode.HorizontalParent("item-2", LayoutSize.StretchedHorizontally(20), new LayoutStyle(new Point(2, 3), 5),
                        LayoutNode.Leaf("item-2a", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2b", LayoutSize.StretchedHorizontally(10)),
                        LayoutNode.Leaf("item-2c", LayoutSize.StretchedHorizontally(10))
                    ),
                LayoutNode.VerticalParent("item-3", LayoutSize.StretchedBoth(), new LayoutStyle(new Point(5, 2), 1),
                        LayoutNode.Leaf("item-3a", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3b", LayoutSize.StretchedBoth()),
                        LayoutNode.Leaf("item-3c", LayoutSize.StretchedHorizontally(10))
                    )
            );

            for (int i = 0; i < 1000; i++)
            {
                var resizedLayout = layout.GetResized(new Point(i * 10, i * 10));
                resizedLayout.Bake();
            }
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void spacer_test()
        {
            var layout = LayoutNode.HorizontalParent("root", LayoutSize.Pixels(50, 20), LayoutStyle.Empty,
                LayoutNode.StretchedSpacer(),
                LayoutNode.Leaf("nudged-item", LayoutSize.StretchedVertically(10)),
                LayoutNode.Spacer(5)
            );

            var firstBakeResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(firstBakeResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void vertical_stretch_test()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 80), LayoutStyle.Empty,
                LayoutNode.HorizontalParent("group-1", LayoutSize.Pixels(50, 20), new LayoutStyle(margin: new Point(3, 3)),
                    LayoutNode.Leaf("tall-item", LayoutSize.StretchedVertically(15)),
                    LayoutNode.Leaf("both-item", LayoutSize.StretchedBoth())
                ),
                LayoutNode.StretchedSpacer(),
                LayoutNode.VerticalParent("group-2", LayoutSize.Pixels(50, 20), new LayoutStyle(margin: new Point(3, 3)),
                    LayoutNode.Leaf("tall-item-2", LayoutSize.StretchedVertically(15)),
                    LayoutNode.Leaf("both-item-2", LayoutSize.StretchedBoth())
                )
            );

            var firstBakeResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(firstBakeResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void create_window_test()
        {
            var headerHeight = 8;
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(80, 40), LayoutStyle.Empty,
                LayoutNode.HorizontalParent("header", LayoutSize.StretchedHorizontally(headerHeight), new LayoutStyle(padding: 2),
                    LayoutNode.StretchedSpacer(),
                    LayoutNode.Leaf("minimize", LayoutSize.Square(headerHeight)),
                    LayoutNode.Leaf("fullscreen", LayoutSize.Square(headerHeight)),
                    LayoutNode.Leaf("close", LayoutSize.Square(headerHeight))
                ),
                LayoutNode.Leaf("canvas", LayoutSize.StretchedBoth())
            );

            var firstBakeResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(firstBakeResult));
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void fixed_aspect_no_spacers()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(36, 64), LayoutStyle.Empty,
                LayoutNode.HorizontalParent("fixed-aspect", LayoutSize.FixedAspectRatio(16, 9), new LayoutStyle(margin: new Point(3, 3)),
                    LayoutNode.Leaf("tall-item", LayoutSize.StretchedVertically(15))
                )
            );

            var firstBakeResult = layout.Bake(); // wide in tall
            var secondBakeResult = layout.GetResized(new Point(64, 36)).Bake(); // wide in wide (perfect match)
            var thirdBakeResult = layout.GetResized(new Point(100, 36)).Bake();

            Approvals.Verify(
                LayoutNodeUtils.DrawResult(firstBakeResult)
                + "\n\n" +
                LayoutNodeUtils.DrawResult(secondBakeResult)
                + "\n\n" +
                LayoutNodeUtils.DrawResult(thirdBakeResult)
                );
        }

        [Fact]
        [UseReporter(typeof(DiffReporter))]
        public void fixed_aspect_with_spacers()
        {
            var layout = LayoutNode.VerticalParent("root", LayoutSize.Pixels(50, 80), LayoutStyle.Empty,
                LayoutNode.StretchedSpacer(),
                LayoutNode.HorizontalParent("aligner", LayoutSize.StretchedBoth(), LayoutStyle.Empty,
                    LayoutNode.StretchedSpacer(),
                    LayoutNode.Leaf("fixed-aspect", LayoutSize.FixedAspectRatio(16, 9)),
                    LayoutNode.StretchedSpacer()
                    ),
                LayoutNode.StretchedSpacer()
            );

            var firstBakeResult = layout.Bake();

            Approvals.Verify(LayoutNodeUtils.DrawResult(firstBakeResult));
        }
    }
}