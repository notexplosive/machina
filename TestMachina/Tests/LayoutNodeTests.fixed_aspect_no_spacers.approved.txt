111111111111111111111111111111111111
1fixed-aspect                      1
1                                  1
1  222222222222222                 1 this gold is wrong (see below)
1  2tall-item    2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  2             2                 1
1  222222222222222                 1
1                                  1
1                                  1
111111111111111111111111111111111111
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
0                                  0
000000000000000000000000000000000000


1111111111111111111111111111111111111111111111111111111111111111
1fixed-aspect                                                  1
1                                                              1
1  222222222222222                                             1
1  2tall-item    2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  2             2                                             1
1  222222222222222                                             1
1                                                              1
1                                                              1
1111111111111111111111111111111111111111111111111111111111111111


1111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
1fixed-aspect                                                                                      1
1                                                                                                  1
1  222222222222222                                                                                 1
1  2tall-item    2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1  2             2                                                                                 1
1002000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000001

185 chars were rendered offscreen and were skipped
`1` at {X:0 Y:36}
`1` at {X:0 Y:37}
`1` at {X:0 Y:38}
`1` at {X:0 Y:39}
`1` at {X:0 Y:40}
`1` at {X:0 Y:41}
`1` at {X:0 Y:42}
`1` at {X:0 Y:43}
`1` at {X:0 Y:44}
`1` at {X:0 Y:45}
`1` at {X:0 Y:46}
`1` at {X:0 Y:47}
`1` at {X:0 Y:48}
`1` at {X:0 Y:49}
`1` at {X:0 Y:50}
`1` at {X:0 Y:51}
`1` at {X:0 Y:52}
`1` at {X:0 Y:53}
`1` at {X:0 Y:54}
`1` at {X:0 Y:55}
`1` at {X:1 Y:55}
`1` at {X:2 Y:55}
`1` at {X:3 Y:55}
`1` at {X:4 Y:55}
`1` at {X:5 Y:55}
`1` at {X:6 Y:55}
`1` at {X:7 Y:55}
`1` at {X:8 Y:55}
`1` at {X:9 Y:55}
`1` at {X:10 Y:55}
`1` at {X:11 Y:55}
`1` at {X:12 Y:55}
`1` at {X:13 Y:55}
`1` at {X:14 Y:55}
`1` at {X:15 Y:55}
`1` at {X:16 Y:55}
`1` at {X:17 Y:55}
`1` at {X:18 Y:55}
`1` at {X:19 Y:55}
`1` at {X:20 Y:55}
`1` at {X:21 Y:55}
`1` at {X:22 Y:55}
`1` at {X:23 Y:55}
`1` at {X:24 Y:55}
`1` at {X:25 Y:55}
`1` at {X:26 Y:55}
`1` at {X:27 Y:55}
`1` at {X:28 Y:55}
`1` at {X:29 Y:55}
`1` at {X:30 Y:55}
`1` at {X:31 Y:55}
`1` at {X:32 Y:55}
`1` at {X:33 Y:55}
`1` at {X:34 Y:55}
`1` at {X:35 Y:55}
`1` at {X:36 Y:55}
`1` at {X:37 Y:55}
`1` at {X:38 Y:55}
`1` at {X:39 Y:55}
`1` at {X:40 Y:55}
`1` at {X:41 Y:55}
`1` at {X:42 Y:55}
`1` at {X:43 Y:55}
`1` at {X:44 Y:55}
`1` at {X:45 Y:55}
`1` at {X:46 Y:55}
`1` at {X:47 Y:55}
`1` at {X:48 Y:55}
`1` at {X:49 Y:55}
`1` at {X:50 Y:55}
`1` at {X:51 Y:55}
`1` at {X:52 Y:55}
`1` at {X:53 Y:55}
`1` at {X:54 Y:55}
`1` at {X:55 Y:55}
`1` at {X:56 Y:55}
`1` at {X:57 Y:55}
`1` at {X:58 Y:55}
`1` at {X:59 Y:55}
`1` at {X:60 Y:55}
`1` at {X:61 Y:55}
`1` at {X:62 Y:55}
`1` at {X:63 Y:55}
`1` at {X:64 Y:55}
`1` at {X:65 Y:55}
`1` at {X:66 Y:55}
`1` at {X:67 Y:55}
`1` at {X:68 Y:55}
`1` at {X:69 Y:55}
`1` at {X:70 Y:55}
`1` at {X:71 Y:55}
`1` at {X:72 Y:55}
`1` at {X:73 Y:55}
`1` at {X:74 Y:55}
`1` at {X:75 Y:55}
`1` at {X:76 Y:55}
`1` at {X:77 Y:55}
`1` at {X:78 Y:55}
`1` at {X:79 Y:55}
`1` at {X:80 Y:55}
`1` at {X:81 Y:55}
`1` at {X:82 Y:55}
`1` at {X:83 Y:55}
`1` at {X:84 Y:55}
`1` at {X:85 Y:55}
`1` at {X:86 Y:55}
`1` at {X:87 Y:55}
`1` at {X:88 Y:55}
`1` at {X:89 Y:55}
`1` at {X:90 Y:55}
`1` at {X:91 Y:55}
`1` at {X:92 Y:55}
`1` at {X:93 Y:55}
`1` at {X:94 Y:55}
`1` at {X:95 Y:55}
`1` at {X:96 Y:55}
`1` at {X:97 Y:55}
`1` at {X:98 Y:55}
`1` at {X:99 Y:36}
`1` at {X:99 Y:37}
`1` at {X:99 Y:38}
`1` at {X:99 Y:39}
`1` at {X:99 Y:40}
`1` at {X:99 Y:41}
`1` at {X:99 Y:42}
`1` at {X:99 Y:43}
`1` at {X:99 Y:44}
`1` at {X:99 Y:45}
`1` at {X:99 Y:46}
`1` at {X:99 Y:47}
`1` at {X:99 Y:48}
`1` at {X:99 Y:49}
`1` at {X:99 Y:50}
`1` at {X:99 Y:51}
`1` at {X:99 Y:52}
`1` at {X:99 Y:53}
`1` at {X:99 Y:54}
`1` at {X:99 Y:55}
`2` at {X:3 Y:36}
`2` at {X:3 Y:37}
`2` at {X:3 Y:38}
`2` at {X:3 Y:39}
`2` at {X:3 Y:40}
`2` at {X:3 Y:41}
`2` at {X:3 Y:42}
`2` at {X:3 Y:43}
`2` at {X:3 Y:44}
`2` at {X:3 Y:45}
`2` at {X:3 Y:46}
`2` at {X:3 Y:47}
`2` at {X:3 Y:48}
`2` at {X:3 Y:49}
`2` at {X:3 Y:50}
`2` at {X:3 Y:51}
`2` at {X:3 Y:52}
`2` at {X:4 Y:52}
`2` at {X:5 Y:52}
`2` at {X:6 Y:52}
`2` at {X:7 Y:52}
`2` at {X:8 Y:52}
`2` at {X:9 Y:52}
`2` at {X:10 Y:52}
`2` at {X:11 Y:52}
`2` at {X:12 Y:52}
`2` at {X:13 Y:52}
`2` at {X:14 Y:52}
`2` at {X:15 Y:52}
`2` at {X:16 Y:52}
`2` at {X:17 Y:36}
`2` at {X:17 Y:37}
`2` at {X:17 Y:38}
`2` at {X:17 Y:39}
`2` at {X:17 Y:40}
`2` at {X:17 Y:41}
`2` at {X:17 Y:42}
`2` at {X:17 Y:43}
`2` at {X:17 Y:44}
`2` at {X:17 Y:45}
`2` at {X:17 Y:46}
`2` at {X:17 Y:47}
`2` at {X:17 Y:48}
`2` at {X:17 Y:49}
`2` at {X:17 Y:50}
`2` at {X:17 Y:51}
`2` at {X:17 Y:52}

